python/lsst/butler/config.py

```diff
@@ -0,0 +1,320 @@
+import os
+import sys
+import warnings
+import yaml
+
+import lsst.utils
+
+from yaml.representer import Representer
+yaml.add_representer(collections.defaultdict, Representer.represent_dict)
+
+# UserDict and yaml have defined metaclasses and Python 3 does not allow multiple
+# inheritance of classes with distinct metaclasses. We therefore have to
+# create a new baseclass that Config can inherit from. This is because the metaclass
+# syntax differs between versions
+
+class _ConfigMeta(type(collections.UserDict), type(yaml.YAMLObject)):
+    pass
+
+
+class _ConfigBase(collections.UserDict, yaml.YAMLObject, metaclass=_ConfigMeta):
+    pass
+
+
+class Config(_ConfigBase):
+    """Config implements a datatype that is used by Butler for configuration parameters.
+    It is essentially a dict with key/value pairs, including nested dicts (as values). In fact, it can be
+    initialized with a dict. The only caveat is that keys may NOT contain dots ('.'). This is explained next:
+    Config extends the dict api so that hierarchical values may be accessed with dot-delimited notiation.
+    That is, foo.getValue('a.b.c') is the same as foo['a']['b']['c'] is the same as foo['a.b.c'], and either
+    of these syntaxes may be used.
+
+    Storage formats supported:
+    - yaml: read and write is supported.
+    """
+
+    def __init__(self, other=None):
+        """Initialize the Config. Other can be used to initialize the Config in a variety of ways:
+        other (string) Treated as a path to a config file on disk. Must end with '.paf' or '.yaml'.
+        other (Config) Copies the other Config's values into this one.
+        other (dict) Copies the values from the dict into this Config.
+        """
+        collections.UserDict.__init__(self)
+
+        if other is None:
+            return
+
+        if isinstance(other, collections.Mapping):
+            self.update(other)
+        elif isinstance(other, Config):
+            self.data = copy.deepcopy(other.data)
+        elif isinstance(other, str):
+            # if other is a string, assume it is a file path.
+            self.__initFromFile(other)
+        else:
+            # if the config specified by other could not be loaded raise a runtime error.
+            raise RuntimeError("A Config could not be loaded from other:%s" % other)
+
+    def ppprint(self):
+        """helper function for debugging, prints a config out in a readable way in the debugger.
+
+        use: pdb> print myConfigObject.pprint()
+        :return: a prettyprint formatted string representing the config
+        """
+        import pprint
+        return pprint.pformat(self.data, indent=2, width=1)
+
+    def __repr__(self):
+        return self.data.__repr__()
+
+    def __initFromFile(self, path):
+        """Load a file from path. If path is a list, will pick one to use, according to order specified
+        by extensionPreference.
+
+        :param path: string or list of strings, to a persisted config file.
```

Comment:
We have 3 different ways of documenting arguments to methods in this file. We need to pick the numpy version.